.github/workflows/ci.yml

```diff
@@ -126,6 +126,13 @@ jobs:
           restore-keys: |
             ${{ matrix.config.os }}-${{ matrix.config.kind }}-
 
+      - name: Cache dprint plugins
+        uses: actions/cache@v1
+        if: matrix.config.kind == 'lint'
+        with:
+          path: ~/.cache/dprint
+          key: ${{ runner.os }}-dprint-plugins-${{ hashFiles('.dprintrc.json') }}
```

Comment:
I think this should be matrix.config.os actually...

Reply:
The plugins are wasm files? What are these? Can they go in third_party/prebuilt? I don't like this added complexity to the CI...

Reply:
When running the first time it needs to download the .wasm files, compile them for the platform, and save the result. This takes anywhere between 2-10s per plugin. After doing this, plugins load quite quickly in about 5-45ms.

So the cache contains the result of downloading & compiling and enables the quick startup times. If we don't want this here, we could move the .wasm files to third_party and apparently compiling is going to be sped up significantly in the next major release of wasmer (I have yet to test it out though I do have access to it). The plugins themselves are between 3-5mb each.

I think it's less complexity and maintenance in the long run to just keep this here though... It's a one time thing we don't have to worry about after and updating the plugins is easy because the urls just need to be changed in the config file.

Reply:
I'd rather have them in third_party than add another root directory.

Alternatively can we opt out of the wasm business and use prettier for markdown? Not super keen on introducing another wasm runtime just for formatting markdown.

Reply:
I'm not sure what you mean by root directory? The cached compiled plugins are stored in `~/.cache/dprint`.

The specific wasm runtime used is an implementation detail that's hidden in the binary. Regardless of what runtime is chosen internally, some caching would be necessary to improve startup times (again, getting it down to 5-45ms to load a plugin).

All the plugins are wasm modules including the typescript formatting. The CLI has no knowledge of the plugins unless they're specified in the configuration file. This allows upgrading plugins by bumping the version url in the config file rather than having to update the CLI binary and also allows for multiple projects to use different versions of plugins that use the same CLI. Since they're wasm modules, they run sandboxed and are portable across systems.

I will work on [this issue](https://github.com/dprint/dprint/issues/242) to support importing plugins from a file path and the compile times should be reasonable for the CI, but I think storing the plugins in third_party is a higher maintenance cost than what's currently done here.

Reply:
Ok thanks for clarifying. Sounds reasonable.

Sure let's go with this added complication to the CI. Maybe add a comment explaining a bit why this is necessary?

Not sure the `if: matrix.config.kind == 'lint'` helps. Maybe remove that line?

Reply:
Sounds good!

By the way, I was thinking about this more and if I parallelize plugin downloading and compiling internally within dprint (right now it downloads and compiles the plugins one after the other) it actually might not add too much time to the CI. I'll try it out with and without and report back with how long they took. It may be that it's not worth the added complexity here in order to save a few seconds. Work will be done in https://github.com/dprint/dprint/issues/222

Reply:
> Edit: Oh wait, I bet these actions are only running with one core. Anyway, I'll see how it goes.

2 core, 7 GB of RAM, 14 GB of disk: https://docs.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources

Reply:
Investigated. 6s format + 1s cache restore with, 13 seconds without, but without probably has higher variability based on download speeds.

I think I'll just remove this to reduce complexity as adding a few seconds to the lint action isn't so bad. Let me know if I should restore this.